src/mem/cache/tags/indexing_policies/base.hh

```diff
@@ -156,10 +199,26 @@ class BaseIndexingPolicy : public SimObject
      * @param entry The entry.
      * @return the entry's original address.
      */
-    virtual Addr regenerateAddr(const Addr tag, const ReplaceableEntry* entry)
-                                                                    const = 0;
+    virtual Addr regenerateAddr(const KeyType &key,
+                                const ReplaceableEntry* entry) const = 0;
+};
+
+class AddrTypes
+{
+  public:
+    using KeyType = Addr;
+    using Params = BaseIndexingPolicyParams;
 };
 
+using BaseIndexingPolicy = IndexingPolicyTemplate<AddrTypes>;
+template class IndexingPolicyTemplate<AddrTypes>;
+
+static constexpr auto
+genTagExtractor(BaseIndexingPolicy *ip)
```

Comment:
I am curious to why this static function is global, and not a static function in the base indexing policy. Is it due to the increased verbosity of doing the latter approach, or is there some kind of other benefit that makes this a better coding practice?

Reply:
Yeah, ideally we should place it there. the issue of keeping in the templated class is the following:

As of now, even in the TaggedEntry, the extractTag is expecting a Addr data type. So we cannot simply add the following

```
static constexpr auto                    
genTagExtractor(IndexingPolicyTemplate *ip)                                                                                                                    
{                                      
    return [ip] (const KeyType &key) { return ip->extractTag(key); };
}  
```

So I went for this quick and dirty option because placing

```
static constexpr auto                    
genTagExtractor(IndexingPolicyTemplate *ip)                                                                                                                    
{                                      
    return [ip] (Addr addr) { return ip->extractTag(addr); };
}  
```

in the template felt equally wrong (it will be defined for non-Addr tag policies as well). But anyway we already have an Addr based extractTag in the template so if you want I can go for the second option.

I believe the cleanest solution would be to add an extra level in the hierarchy and move tag extraction logic over there so that it does not reside in the template. In lots of cases you don't really need to extract the Tag and you shouldn't be doing it for structured tags

Reply:
Nah, it is fine to leave as is; I was just wondering about the thought process.